### ISSUE TITLE
Having the option to enter custom commands & run trainers in WHDLoad hdf-files?

### ISSUE THREAD

**Issue by Gagert** [libretro/libretro-uae#179]
Being a former user of fs-uae before fully migrating into this libretro core for my Amiga emulation needs, there is one thing I'm missing from fs-uae that as far as I know is missing in libretro uae:

The options of entering custom commands and use the bundled trainers that comes with some of the WHDLoad Slaves to some of the Amiga games.

For instance, The Settlers has the ability to skip the intro by entering Custom1 as a launch-command before starting the WHDLoad. Some other games have optional trainers at the very start of the games, in other words when the AmigaDOS screen is in the background and the tiny box in the middle has a description with the name of the game you're running, before the actual game starts (in this box the trainers are often placed by having various toggle-able checkboxes, and info like type Custom1 to skip intro is displayed here).

There are however only a handful of games where I personally miss the introduction screen before the game actually starts (because of trainers and relevant info), other than that I think it's fine the way it is now. Maybe having a toggleable option for this introduction screen in this core's Options menu?

**Comment by sonninnos** [comment]
I have no idea how FS-UAE does that, and I did try to search it in the source, but could not find anything useful. Is it a separate launcher of some sort..?

Anyway, doing anything like that corewise would be much more complicated than just doing it manually by editing the global `S:WHDLoad.prefs` and/or the game icon via Workbench. At least the global prefs has `ButtonWait`, `ConfigDelay` and `SplashDelay`, but `Custom`s need to be edited per game of course.

I also haven't checked what the "official" Whdload.hdf does at startup, because I use my own method of launching through CLI, which is a manually made script per game that simply does `whdload Settlers.Slave preload custom1=1`.

But for now I can't see it at all as a job for the core to interject those configurations, nor could I vision how to accomplish it cleanly.

**Comment by Gagert** [comment]
Could it be a possibility to make it so that you could write in the custom1 stuff inside a m3u-file itself (a m3u-file that launches the hdf-file with custom-parameters), or even a parenthesis after the filename of the amiga hdf-file that gets read by this libretro core itself (like having "The Settlers (custom1=1).hdf" to skip the intro in this WHDLoad)? I don't know enough about programming regarding how much work it would be to make the core somehow read text outside the core (filenames and text inside m3u-files), and apply it to the hdf-file itself upon launching.

Regarding the technicalities of FS-UAE while launching WHDLoad-files, I'm not quite sure if it's using a separate launcher of some sort when launching WHDLoads. I imagined a proper configured Amiga 1200 with extra memory installed & harddrive would launch games the same way, if configured the right way (if the Amiga is configured to quicklaunch the game without starting Workbench properly)?

**Comment by rsn8887** [comment]
Have you tried just commenting out or deleting the line
ConfigDelay=0
in your WHDLoad.prefs file?

AFAIK, that line suppresses the screen that shows the skip intro buttons and other custom option during launch.

**Comment by sonninnos** [comment]
Custom commands like that would have to be ninja edited on the fly to the startup script command somehow. I can't promise anything, but I'll look into it. 

Also maybe that `WHDLoad.prefs` could be made to be a RA system file alongside the whdload.hdf, which would override the file inside the .hdf, so editing it would be more user friendly. There aren't too many easy to use Amiga-side file editors by default (Ed is clumsy and Edit is horrible. I like MicroEMACS the most).

But for now the only option is to manually edit the file `S:WHDLoad.prefs` to make `ConfigDelay` a reasonable number to have enough time to move the mouse (why on earth is there not a separate start-button after doing the configs, because now it just starts even if you are ready or not) and the checkbox screen will appear, if the game supports it.

Looks like Settlers does not, it just displays that the custom option exists. Now there is no way of including `Custom1=1`, because the whdload.hdf just launches game.slave with `Preload`, so editing the icon's tooltypes won't do anything.

Midnight Resistance for instance offers a huge list of options.

Only thing I can think of for those cases is to make a game specific load script inside the game .hdf and edit the startup-sequence in whdload.hdf to check if the script exists, and execute it instead of the default method.

Like so:
- Make a file to the game .hdf that has `whdload Settlers.Slave preload custom1=1` and name it `load` for instance. Scripts need to have the script flag with `protect load +s`, but the whdload.hdf of course does not have that command, nor any editor for that matter.
- Edit the `S:startup-sequence` to do:
```
if exists dh1:load
   load
else
   whdload game.slave preload
endif
```

That hdf guide sure is very outdated regarding the core. Could someone advise him to update it..?

**Comment by sonninnos** [comment]
All right, I have a new WHDload.hdf ready for action with the following changes:
- No need to rename slave to game.slave
- Ability to load custom parameters by either creating a file called `custom` in the game .hdf which only has the exceptions, like `buttonwait custom1=1 configdelay=-1 splashdelay=200` etc OR a file called `load` which is the whole whdload command
-- If either of those are not found, the slave is launched with just the default `preload` like it is now
- Added Devs:Monitors/NTSC, so NTSC parameter can be used
- Added ClickNot
- Removed useless Kickstart rtb's

The image size could be easily halved, and there is still plenty of space for the saves, which are still redirected there instead of the game directory. With these changes it should be a real possibility to also use .lha's directly as hard drives, if and when I can get it working properly.

Just yell if you'd like to take the new .hdf for a test drive before I do a PR. I'll investigate next if those custom commands and WHDLoad defaults could be inserted without modifying the .hdfs.
